quickwit/quickwit-config/src/config.rs

```diff
@@ -190,36 +206,47 @@ impl Default for SearcherConfig {
     }
 }
 
-#[derive(Derivative)]
-#[derivative(Debug)]
-#[derive(Clone, Serialize, Deserialize, PartialEq)]
+#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq)]
+struct List(Vec<String>);
+
+impl FromStr for List {
+    type Err = anyhow::Error;
+
+    fn from_str(list_str: &str) -> Result<Self, Self::Err> {
+        let list = list_str
+            .split(',')
+            .map(|elem| elem.trim().to_string())
+            .filter(|elem| !elem.is_empty())
+            .collect();
+        Ok(List(list))
+    }
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
 #[serde(deny_unknown_fields)]
 struct QuickwitConfigBuilder {
-    version: usize,
+    #[serde(default)]
+    version: ConfigValue<usize, QW_NONE>,
     #[serde(default = "default_cluster_id")]
-    cluster_id: String,
+    cluster_id: ConfigValue<String, QW_CLUSTER_ID>,
     #[serde(default = "default_node_id")]
-    node_id: String,
+    node_id: ConfigValue<String, QW_NODE_ID>,
+    #[serde(default = "default_enabled_services")]
+    enabled_services: ConfigValue<List, QW_ENABLED_SERVICES>,
     #[serde(default = "default_listen_address")]
-    listen_address: String,
-    advertise_address: Option<String>,
+    listen_address: ConfigValue<String, QW_LISTEN_ADDRESS>,
+    advertise_address: ConfigValue<String, QW_ADVERTISE_ADDRESS>,
     #[serde(default = "default_rest_listen_port")]
-    rest_listen_port: u16,
-    #[serde(default = "default_quickwit_services")]
-    enabled_services: Vec<String>,
-    gossip_listen_port: Option<u16>,
-    grpc_listen_port: Option<u16>,
-    #[serde(default)]
-    peer_seeds: Vec<String>,
+    rest_listen_port: ConfigValue<u16, QW_REST_LISTEN_PORT>,
+    gossip_listen_port: ConfigValue<u16, QW_GOSSIP_LISTEN_PORT>,
+    grpc_listen_port: ConfigValue<u16, QW_GRPC_LISTEN_PORT>,
     #[serde(default)]
-    #[derivative(Debug(format_with = "redact_uri"))]
-    metastore_uri: Option<String>,
-    #[serde(default)]
-    #[derivative(Debug(format_with = "redact_uri"))]
-    default_index_root_uri: Option<String>,
-    #[serde(default = "default_data_dir_path")]
+    peer_seeds: ConfigValue<List, QW_PEER_SEEDS>,
     #[serde(rename = "data_dir")]
-    data_dir_path: PathBuf,
+    #[serde(default = "default_data_dir_uri")]
+    data_dir_uri: ConfigValue<Uri, QW_DATA_DIR>,
```

Comment:
Not sure to understand the change to `Uri`, is it justified for windows compatibility?

Reply:
This simplifies the code in a few places in the config builder. The config remains with `data_dir_path: PathBuf`.